### PR TITLE
feat: Add API for experiment list columns [WEB-979]

### PR DESCRIPTION
**Pull request by gt2345**
## Description

Add GET `/api/v1/projects/{project_id}/columns` to return all available columns for the column picker (and Filter, Sort, Group dropdowns) on a per project basis.

Sample response:
```
{
    "general": [
        "GENERAL_COLUMN_ID",
        "GENERAL_COLUMN_DESCRIPTION",
        "GENERAL_COLUMN_USER",
        "GENERAL_COLUMN_STARTTIME",
        "GENERAL_COLUMN_UNSPECIFIED",
        "GENERAL_COLUMN_NAME",
        "GENERAL_COLUMN_TAGS",
        "GENERAL_COLUMN_DURATION",
        "GENERAL_COLUMN_STATE"
    ],
    "hyperparameters": [
        "classifier.learning_rates",
        "self_supervised.lars_eta",
        "classifier.logit_clipping.alpha",
        "dropout1",
        "dropout2",
        "gain_per_batch",
        "global_batch_size",
        "training_structure.error",
        "backbone_name",
        "classifier.train_epochs",
        "classifier.logit_regularization_beta",
        "n_filters2",
        "metrics_sigma",
        "layer1_dropout",
        "training_mode",
        "self_supervised.learning_rate.base",
        "starting_base_value",
        "learning_rate",
        "self_supervised.learning_rate.warmup_epochs",
        "classifier.momentum",
        "training_structure.more_loss",
        "validation_structure.more_error",
        "learning_rate_decay",
        "self_supervised.learning_rate.base_batch_size",
        "layer3_dropout"
    ],
    "metrics": [
        "validation.accuracy",
        "validation.validation_loss"
    ]
}
```

Benchmark for migrations:
- aggregate hyper parameters by project
           6.5k experiments 19 projects
           result: 640ms
- aggregate validation metrics by experiment
           6.5k experiments 41k trials 82k raw_validations
           result: 818ms

## Test Plan

Create new experiment within a project, verify that hyperparameters and metrics names are in the API call return.
Move an experiment to a project, verify that hyperparameters and metrics names are in the API call return for the destination project.
Verify that code tab and download experiment code function as expected.

## Commentary (optional)

Hyper parameters and metrics name aggregation do not consider if the instance has been achieved or not.
Validation metrics name aggregation only consider `best_validation` for performance reason. 

## Checklist

- [x] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

WEB-979

### CODE REVIEW

File: master/internal/api_project.go [hunk 1]
```diff
@@ -37,6 +39,87 @@ func (a *apiServer) GetProjectByID(
...
+	err = db.Bun().
+		NewSelect().Table("exp_metrics_name").ColumnExpr(
+		"json_array_elements_text(vname) AS name").Distinct().Where(
+		"project_id = ?", id).Scan(ctx, &metricNames)
+	if err != nil {
+		return nil, err
+	}
+	for _, mn := range metricNames {
+		metricColumns = append(metricColumns, fmt.Sprintf("%s.%s", "validation", mn.Name))
+	}
```
Comment (ashtonG) [inline master/internal/api_project.go:116]: Could we combine these queries, either by doing a union or by discarding duplicate values ourselves after running the query?

File: proto/src/determined/project/v1/project.proto [hunk 2]
```diff
@@ -10,6 +10,28 @@ import "google/protobuf/timestamp.proto";
...
+  GENERAL_COLUMN_TAGS = 4;
+  // Column user.
+  GENERAL_COLUMN_USER = 5;
+  // Column starttime.
+  GENERAL_COLUMN_STARTTIME = 6;
+  // Column duration.
+  GENERAL_COLUMN_DURATION = 7;
+  // Column state.
+  GENERAL_COLUMN_STATE = 8;
+}
```
Comment (hkang1) [inline proto/src/determined/project/v1/project.proto:33]: **question:** There might be a couple of general options missing. Lookin at the UI:
- Resource Pool
- Progress
- Forked
- Searcher Type
- Trial Count
- Checkpoint Size
- Checkpoint Count
Screenshot 2023-03-22 at 9 36 05 AM

**Review by hkang1 — COMMENTED** [review]
I didn't look through all the go code, but the hyperparameters (nested, nice!) and the metric ids look awesome. In the general columns there might be some missing? Noted it below.

File: proto/src/determined/project/v1/project.proto (hunk 2 above)
Comment (gt2345) [reply proto/src/determined/project/v1/project.proto:33]: I've only included the ones presented in the figma, I thought eliminating some columns was also part of design, but I can definitely add them back.

File: proto/src/determined/project/v1/project.proto [hunk 3]
```diff
@@ -10,6 +10,28 @@ import "google/protobuf/timestamp.proto";
 import "google/protobuf/wrappers.proto";
 import "protoc-gen-swagger/options/annotations.proto";
 
+// GeneralColumn is the pre-defined column names for experiment list table.
+enum GeneralColumn {
```
Comment (ioga) [inline proto/src/determined/project/v1/project.proto:14]: I am not sure this enum is worth it. Maybe we could have simple strings, just like hyperparameters and metrics? 
- it'd make the API look more uniform
- we can validate if the strings are in an allowlist of columns in go code instead

File: master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  INSERT INTO exp_hyperparameters (project_id, experiment_id, hyperparameters) VALUES (
+    NEW.project_id, NEW.id, json_build_array((NEW.config->'hyperparameters'))
+  ) ON CONFLICT(experiment_id) DO UPDATE SET hyperparameters = EXCLUDED.hyperparameters;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER autoupdate_exp_hyperparameters
+AFTER INSERT OR UPDATE ON experiments
+FOR EACH ROW EXECUTE PROCEDURE autoupdate_exp_hyperparameters();
```
Comment (ioga) [inline master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql:26]: why not have this as a column on the `experiments` table?

File: master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        WHERE  t.experiment_id = (SELECT experiment_id FROM trials WHERE id = NEW.trial_id) AND t.id = v.trial_id AND e.id=t.experiment_id GROUP BY (e.project_id, e.id)
+    )  ON CONFLICT(experiment_id) DO UPDATE SET vname = EXCLUDED.vname;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER autoupdate_exp_validation_metrics_name
+AFTER INSERT ON raw_validations
+FOR EACH ROW EXECUTE PROCEDURE autoupdate_exp_validation_metrics_name();
+
```
Comment (ioga) [inline master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql:93]: I thought we were not going to include metrics here until summary metrics are done: how would these metrics columns be useful if we can't do any filtering/sorts using them? cc @hkang1 

There's a work-in-progress proposal for summary metrics structure [in a "Summary Metrics" section here](https://hpe.sharepoint.com/:w:/r/teams/detai/Shared%20Documents/Engineering/Backend/Unified%20metrics/ERD-%20Unified%20Metrics.docx?d=w6ff07466cd42461593056a8aae14a7e6&csf=1&web=1&e=hfsoed). it'll effectively supercede this table.

**Review by ioga — CHANGES_REQUESTED** [review]
(noticed this is approved, so will "request changes" instead)

File: master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql (hunk 4 above)
Comment (ioga) [reply master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql:26]: also, do we even needs this?.. is the `config->'hyperparameters'` scan really that much slower?

File: proto/src/determined/project/v1/project.proto (hunk 3 above)
Comment (gt2345) [reply proto/src/determined/project/v1/project.proto:14]: I added these in proto because I want this to be a shared information between front end and back end, so front end would know this information before API response. I wonder what would be the disadvantage against using Enum?

File: master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql (hunk 4 above)
Comment (gt2345) [reply master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql]: `exp_hyperparameters` table removed

File: proto/src/determined/project/v1/project.proto (hunk 3 above)
Comment (ioga) [reply proto/src/determined/project/v1/project.proto:14]: with enums, query parameters will look like `?columns=GENERAL_COLUMN_ID&columns=GENERAL_COLUMN_NAME&hyperparameters=dropout1&hyperparameters=lr&metrics=training.loss`
without, `?columns=id&columns=name&hyperparameters=dropout1&hyperparameters=lr&metrics=training.loss`

File: master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql (hunk 5 above)
Comment (hkang1) [reply master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql:64]: @ioga does summary metrics include validation metrics? I thought the columns should still at least return something like:
```
metrics: [
  'validation.accuracy',
  'validation.loss',
],
```
and it would exclude things like `training.accuracy` and `training.loss` until summary metrics work is done.

File: master/static/srv/move_experiment.sql [hunk 6]
```diff
@@ -1,3 +1,43 @@
-UPDATE experiments SET project_id = $2
-WHERE id = $1
-RETURNING id;
+WITH update_mn AS (
```
Comment (ioga) [inline master/static/srv/move_experiment.sql:1]: I'd rather call `updateProjectHyperparameters` in the same transaction as the old query. this'd be much better for the code readability.

File: master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql (hunk 5 above)
Comment (ioga) [reply master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql:82]: today there's only one summary metric: best searcher metric, which is a validation metric. other validation metrics aren't included.

based on my chat with Guangqing, these columns will be used in place of existing `MetricNames` to get the column names for *time series* requests. if these column names are intended for filtering, we need to do this differently.

File: proto/src/determined/api/v1/api.proto [hunk 7]
```diff
@@ -1728,6 +1728,16 @@ service Determined {
...
     };
   }
+  // Get a list of columns for experiment list table.
+  rpc GetProjectColumns(GetProjectColumnsRequest)
+      returns (GetProjectColumnsResponse) {
+    option (google.api.http) = {
+      get: "/api/v1/projects/{id}/columns"
+    };
+    option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
+      tags: "Projects"
```
Comment (hamidzr) [inline proto/src/determined/api/v1/api.proto:1738]: should we call this an internal API to give us more flexibility? or do we see it as something users might externally use?

File: master/internal/api_project.go [hunk 8]
```diff
@@ -37,6 +38,69 @@ func (a *apiServer) GetProjectByID(
...
+		TableExpr("LATERAL json_array_elements_text(tname) AS tnames").
+		ColumnExpr("array_to_json(array_agg(DISTINCT tnames)) AS tname").
+		ColumnExpr("array_to_json(array_agg(DISTINCT vnames)) AS vname").
+		Where("project_id = ?", id).Scan(ctx, &metricNames)
+	if err != nil {
+		return nil, err
+	}
+	for _, mn := range metricNames {
+		for _, mnt := range mn.Tname {
+			metricColumns = append(metricColumns, fmt.Sprintf("%s.%s", "training", mnt))
```
Comment (hamidzr) [inline master/internal/api_project.go:92]: question: what do we get out of indexing the metric names and is `.` a standard notation for us? is it going to be parsed? 

```
  "metrics": [
    "training.loss1",
    "training.loss2",
    "validation.loss1",
    "validation.loss2"
  ]
  ```

File: proto/src/determined/project/v1/project.proto [hunk 9]
```diff
@@ -10,6 +10,42 @@ import "google/protobuf/timestamp.proto";
...
+  // Unspecified column.
+  GENERAL_COLUMN_UNSPECIFIED = 0;
+  // Column id.
+  GENERAL_COLUMN_ID = 1;
+  // Column name.
+  GENERAL_COLUMN_NAME = 2;
+  // Column discription.
```
Comment (hamidzr) [inline proto/src/determined/project/v1/project.proto:21]: typo

**Review by hamidzr — COMMENTED** [review]
I reviewed API-level changes.

File: master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql (hunk 5 above)
Comment (gt2345) [reply master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql:82]: I believe filtering functionality will come after the unified metrics work, right now we just need these metrics name in place of `MetricNames` call

File: master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+FOR EACH ROW EXECUTE PROCEDURE autoupdate_exp_training_metrics_name();
+
+CREATE OR REPLACE FUNCTION autoupdate_exp_validation_metrics_name() RETURNS trigger AS $$
+BEGIN
+    INSERT INTO exp_metrics_name (project_id, experiment_id, vname) (
+        SELECT e.project_id, e.id, array_to_json(array_agg(DISTINCT names)) AS vname
+        FROM experiments e, trials t, raw_validations v, LATERAL jsonb_object_keys(v.metrics->'validation_metrics') AS names
+        WHERE  t.experiment_id = (SELECT experiment_id FROM trials WHERE id = NEW.trial_id) AND t.id = v.trial_id AND e.id=t.experiment_id GROUP BY (e.project_id, e.id)
+    )  ON CONFLICT(experiment_id) DO UPDATE SET vname = EXCLUDED.vname;
+  RETURN NEW;
```
Comment (ioga) [inline master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql:75]: I don't think this approach will be fast enough. doing a full scan on steps table on every insert will be very expensive for long running experiments. we need to update the aggregate by only using the existing aggregate and the new value.

File: master/static/srv/move_experiment.sql (hunk 6 above)
Comment (gt2345) [reply master/static/srv/move_experiment.sql]: This has been fixed

File: proto/src/determined/api/v1/api.proto (hunk 7 above)
Comment (gt2345) [reply proto/src/determined/api/v1/api.proto]: I think this should be an internal API for now

File: master/internal/db/postgres.go [hunk 11]
```diff
@@ -249,20 +249,20 @@ func (db *PgDB) namedExecOne(query string, arg interface{}) error {
...
+// 	}
+// 	if sErr := nstmt.QueryRowx(arg).Scan(dest); sErr != nil {
+// 		err = errors.Wrapf(sErr, "error scanning query %s", query)
+// 	}
+// 	if cErr := nstmt.Close(); cErr != nil && err != nil {
+// 		err = errors.Wrap(cErr, "error closing named DB statement")
+// 	}
+
+// 	return err
+// }
```
Comment (ioga) [inline master/internal/db/postgres.go:265]: remove it.

**Review by ioga — APPROVED** [review]

File: master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        SELECT e.project_id, e.id, mname.mdata AS vname
+        FROM experiments e, trials t, mname
+        WHERE t.id = NEW.trial_id AND e.id = t.experiment_id 
+    )  ON CONFLICT(experiment_id) DO UPDATE SET vname = EXCLUDED.vname;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER autoupdate_exp_validation_metrics_name
+AFTER INSERT ON raw_validations
```
[the remark this answers is not in the document]
Comment (gt2345) [reply master/static/migrations/20230315150937_add-hyperparameters-metrics-name-view.tx.up.sql:62]: I kinda want to keep it this way, suppose an experiment intentionally add a specific validation metrics name, I do want it to show up, since it should be okay performance wise. 
Using `best_validation_id` during migration is mainly for performance reason, with unified metrics work coming up, I don't want users to go through two long migrations.
